wiki/Client Security.md

```diff
@@ -0,0 +1,95 @@
+
+After comparing several alternatives, we believe [argon2](https://github.com/P-H-C/phc-winner-argon2) is the optimal choice to replace step (2) of **GenLeaf** in the next version of ONE Wallet. Similar to scrypt, it can be configured to use a large size of memory and impose an arbitary workload to slow down the computation. Aside for its multiple advantages over other candidates (which are discussed at length in their Github repository, the password-hashing competition, and their paper), here are the other factors we have considered:
+
+- Argon2 has a [browser-based implementation](https://github.com/antelle/argon2-browser) based on WebAssembly. It offers satisfactory performance under reasonably low memory settings, and can complete within a reasonable amount of time for 10<sup>6</sup> repeatitions. Therefore, it would not slow down the setup process
+- Based on our knowledge, no one has built an ASIC miner for computing Argon2 hashes at this time. 
+- Argon2 is not yet even popular enough to be included in [hashcat](https://hashcat.net/hashcat/), the well-known GPU/FPGA based password recovery tool by brute-forcing hash computations via GPU.
+
+<a name="f1">[1]</a> See an outdated [Litecoin Mining Hardware Comparison](https://litecoin.info/index.php/Mining_hardware_comparison)
+
+<a name="f2">[2]</a> See [AntMiner L3+](https://shop.bitmain.com/promote/antminer_l3_litecoin_asic_scrypt_miner/specification) which achieves 504MH/s
+
+#### III. Scrambled Memory Layout
+
+We can further improve the security of **GenLeaf** by placing the result in the memory where it would be difficult for the attacker to locate, but would be easy for the client with the correct OTP to retrieve. A naive construct following this intuition is to store each leaf at the position of the OTP they correspond to and a pointer to their neighbor. Unfortunately this naive approach leaks the OTP code of each leaf and renders OTP useless. A mild improvement from the naive approach is to swap each leaf with its neighbor's position (after first repositioning all leaves to their OTP values) and remove the pointer to the neighbor (since neighbor is part of the Merkle Proof, not the leaf node itself). This construct no longer leaks the OTP for each leaf, but an attacker would still be unable to identify the location for each pair of leaves by pre-comptuing the parent hash of all pairs of leaves, which can be done in M<sup>2</sup> operations and would only require 32 bytes * M amount of memory. Since each operation is a SHA256 hash, based on the analysis in Section I, an attacker with a high-end GPU only needs ~10 minutes to identify exact positions of all leaves before they are scrambled. Moreover, we cannot use the techniques in Section II by replacing the hash function to a complex one, since the hash function must be supported on blockchain as well and must be economical and fast enough to compute on blockchain.
```

Comment:
From @ivan-homoliak-sutd :

> But Merkle proof is computed at the client. So, with scrambled design, I do not see how you can compute it since you do not know where are the other leaves required to compute the proof. See also below. (This is similar even in the improved version.)

> I understood that you store only leafs and not intermediary nodes of Merkle tree at the client. However,  it is also convenient to store intermediary nodes, so you can construct proofs at the client faster - just by accessing the continuous array of leafes and intermediary nodes (no need to do Merkle tree aggregation/reduction). This holds only if you do not have scrambled memory. With scrambled memory, this might be a problem. Anyway, I do not understand how the client constructs proofs even with scrambled design. To do so, the client needs also the knowledge of positions of other leafs (additionally intermediary nodes in the tree), while these positions can be obtained only when other OTPs are provided (which is unrealistic). Could you please clarify?

Reply:
The entire tree is stored at the client, not just the leaves. I don't think there is much benefit of storing only the leaves since it would only reduce storage consumption by 50%.

Reply:
Yes, I agree to store the entire tree. 

"the number of operations required for simple brute-force is increased to (4M)^8" 
=> 4M does not seem correct. When you generate the 4 random numbers by PRNG, you can put the range of even 2^256 (by using hash) instead and store the elements in the mapping, and according to your formula, it'd mean unbelievable (2^256M)^8. However, the entropy of one OTP is only 2^20 ~ 2.5B, which, to me, is also manifested to PRNG you use. So, in fact, you have only 1M options => (1M)^8 ~ 160bits; which is pretty good.

Assuming that the left sibling was already submitted, this degrades to (1M)^4 ~ 80bits, which is still strong enough to be brute-forced in 30s or 60s (until the expiration of the right sibling).

**Some minor comments:**

Also, can you please write that by the "result", you mean leaves (further containing controlled randomness)?

"...attacker would still be **unable** to identify the location" => able

"Moreover, we cannot use the techniques in Section II by replacing the hash function to a complex one, since the hash 
function must be supported on blockchain" 
=> The complex hash function would be used only for computation of EOTP at the client, which is agnostic to the blockchain; so it can be used.

"to generate 4 **consecutive** numbers between [1, 4M]" 
=> why consecutive? I'd say random.

Reply:
1. Yes, **unable** is a typo and should be replaced with **able**.
2. We cannot use the complex hash function because the brute-force operation being considered here is the one that hashes two leaf nodes into a parent node. The attacker only needs to recover the correct position of leaf nodes here from the scrambled memory layout. After correct positions are identified, the attacker can proceed to brute-force one or more individual leaf nodes, which the complexity is only additive, not multiplicative , to the overall cost.
3. Yes, `Result` as used in the opening sentence of this line `We can further improve the security of **GenLeaf** by placing the result in the memory` means a leaf (not leaves, since **GenLeaf** only generates one at a time).
4. The comment about consecutive numbers seem to be referring to a different line. Let's open another thread there because there are too many comments in this line.
5.  Similarly, the comment about number of operations required for simple brute-force is on a separate line. I will copy your comment and start a thread there.